src/wasm/mod.rs

```diff
@@ -0,0 +1,358 @@
+    let writer = std::io::BufWriter::new(fingerprint_file);
+    serde_json::to_writer_pretty(writer, &fingerprints)?;
+    Ok(())
+}
+
+/// Compiles a WASM module located at `input` and writes an optimized shared object to `output`.
+fn compile(
+    input: impl AsRef<Path> + Debug,
+    output: impl AsRef<Path> + Debug,
+) -> Result<Fingerprint> {
+    let input = input.as_ref();
+    let fingerprint = calculate_fingerprint(input)?;
+
+    let mut bindings = lucet_wasi::bindings();
+    bindings.extend(&Bindings::from_str(include_str!("hostcall/bindings.json"))?)?;
+    Lucetc::new(input)
+        .with_bindings(bindings)
+        .shared_object_file(output)?;
+
+    event!(Level::INFO, "done");
+    Ok(fingerprint)
+}
+
+/// A foreign module that is operating as a WASM guest.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct WasmModule {
+    /// A stored version of the config for later referencing.
+    config: WasmModuleConfig,
+    /// The handle to the Lucet instance.
+    #[derivative(Debug = "ignore")]
+    instance: InstanceHandle,
+    role: Role,
+}
+
+impl WasmModule {
+    /// Build the WASM instance from a given config.
+    pub fn build(config: impl Into<WasmModuleConfig> + Debug) -> Result<Self> {
+        event!(Level::TRACE, "instantiating");
+        let config = config.into();
+        let output_file = config
+            .artifact_cache
+            .join(config.path.file_stem().ok_or("Must load files")?)
+            .with_extension("so");
+
+        // Prepwork
+        fs::create_dir_all(&config.artifact_cache)?;
+        hostcall::ensure_linked();
+
+        let calculated_fingerprint = calculate_fingerprint(&config.path)?;
+        let found_fingerprint = fetch_fingerprint(&config.artifact_cache, &config.path)?;
+        match found_fingerprint {
+            Some(found) if found == calculated_fingerprint => {
+                // We can be lazy and do nothing! How wonderful.
+                // println!("Skipped wasm compile! Make me a metric later!");
+            }
+            None | Some(_) => {
+                // println!("Rebuild wasm! Make me a metric later!");
+                let internal_event_compilation =
+                    internal_events::WasmCompilation::begin(config.role);
+                let fingerprint = compile(&config.path, &output_file)?;
+                store_fingerprint(&config.artifact_cache, &config.path, fingerprint)?;
+                internal_event_compilation.complete();
+            }
+        }
+
+        // load the compiled Lucet module
+        let module = DlModule::load(&output_file).unwrap();
+
+        // create a new memory region with default limits on heap and stack size
+        let region = &MmapRegion::create(
+            1,
+            &Limits {
+                heap_memory_size: config.max_heap_memory_size,
+                ..Limits::default()
+            },
+        )?;
+
+        // instantiate the module in the memory region
+        let instance = region
```

Comment:
I considered trying to make this a shared region between possibly many instances, but left that for later work since we only run transforms on one thread right now.